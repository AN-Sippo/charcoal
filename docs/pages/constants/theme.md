# `@charcoal/theme`

charcoal 全体で制約されるテーマ構造の型を提供します。

## インストール

npm

```bash
npm i @charcoal/theme
```

yarn

```bash
yarn add @charcoal/theme
```
