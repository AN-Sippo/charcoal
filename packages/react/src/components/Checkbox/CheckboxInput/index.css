.charcoal-checkbox-input {
  appearance: none;
  display: flex;
  cursor: pointer;
  margin: 0;
  width: 20px;
  height: 20px;
  border-radius: 4px;
  transition: 0.2s box-shadow, 0.2s background-color;
}

.charcoal-checkbox-input:disabled,
.charcoal-checkbox-input[readonly] {
  opacity: 0.32;
  cursor: default;
}

.charcoal-checkbox-input:checked {
  background-color: var(--charcoal-brand);
}

.charcoal-checkbox-input:checked::after {
  content: '';
  color: white;
  display: flex;
  margin: auto;
  width: 16px;
  height: 16px;
  background-image: url('data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22none%22%3E%3Cpath%20fill%3D%22%23FFFFFF%22%20d%3D%22M15.848%207.449a2%202%200%200%201%202.804%202.853l-8.664%208.514-4.402-4.402a2%202%200%201%201%202.828-2.828l1.598%201.598%205.836-5.735z%22%2F%3E%3C%2Fsvg%3E');
  background-repeat: no-repeat;
  background-size: contain;
  background-position: center;
}

.charcoal-checkbox-input:not(:checked) {
  border-width: 2px;
  border-style: solid;
  border-color: var(--charcoal-text4);
}

.charcoal-checkbox-input:not(:disabled):focus {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-checkbox-input:not(:disabled):focus-visible {
  box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
}
.charcoal-checkbox-input:not(:disabled):focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-checkbox-input:checked:not(:disabled):hover {
  background-color: var(--charcoal-brand-hover);
}
.charcoal-checkbox-input:checked:not(:disabled):active {
  background-color: var(--charcoal-brand-press);
}

.charcoal-checkbox-input[aria-invalid='true'],
.charcoal-checkbox-input[aria-invalid='true']:not(:disabled):focus {
  box-shadow: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
