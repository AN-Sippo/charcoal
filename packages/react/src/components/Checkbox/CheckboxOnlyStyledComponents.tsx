import React, { forwardRef, memo } from 'react'
import styled, { css } from 'styled-components'
import { disabledSelector } from '@charcoal-ui/utils'

import Icon from '../Icon'

type CheckboxLabelProps =
  | {
      children: React.ReactNode
    }
  | {
      label: string
    }

export type CheckboxProps = CheckboxLabelProps & {
  readonly id?: string
  readonly name?: string

  readonly checked?: boolean
  readonly defaultChecked?: boolean
  readonly disabled?: boolean
  readonly readonly?: boolean

  readonly onClick?: () => void
  readonly onChange?: (isSelected: boolean) => void
  readonly onBlur?: () => void
  readonly onFocus?: () => void
}

const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  function CheckboxInner(props, ref) {
    const isDisabled = (props.disabled ?? false) || (props.readonly ?? false)
    return (
      <InputRoot aria-disabled={isDisabled}>
        <CheckboxRoot>
          <CheckboxInput
            type="checkbox"
            aria-label={'label' in props ? props.label : undefined}
            ref={ref}
            checked={props.checked ?? false}
            disabled={props.disabled}
            readOnly={props.readonly}
          />
          <CheckboxInputOverlay
            aria-hidden={true}
            checked={props.checked ?? false}
          >
            <Icon name="24/Check" unsafeNonGuidelineScale={2 / 3} />
          </CheckboxInputOverlay>
        </CheckboxRoot>

        {'children' in props && <InputLabel>{props.children}</InputLabel>}
      </InputRoot>
    )
  }
)

export default memo(Checkbox)

const hiddenCss = css`
  visibility: hidden;
`

const InputRoot = styled.label`
  position: relative;
  display: flex;

  cursor: pointer;
  ${disabledSelector} {
    cursor: default;
    opacity: ${({ theme }) => theme.elementEffect.disabled.opacity};
  }

  gap: 4px;
`

const CheckboxRoot = styled.div`
  position: relative;
`

const CheckboxInput = styled.input`
  &[type='checkbox'] {
    appearance: none;
    display: block;
    cursor: pointer;
    margin: 0;
    width: 20px;
    height: 20px;

    transition: 0.2s box-shadow;

    &:checked {
      background: ${({ theme }) => theme.color.brand};
    }
    &:not(:checked) {
      border-width: 2px;
      border-style: solid;
      border-color: ${({ theme }) => theme.color.text4};
    }
    border-radius: 4px;

    :focus {
      box-shadow: 0 0 0 ${({ theme }) => theme.outline.default.weight}px
        ${({ theme }) => theme.outline.default.color};
    }
  }
`

const CheckboxInputOverlay = styled.div<{ checked?: boolean }>`
  position: absolute;
  top: -2px;
  left: -2px;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;

  width: 24px;
  height: 24px;
  color: ${({ theme }) => theme.color.text5};

  ${({ checked }) => checked !== true && hiddenCss};
`

const InputLabel = styled.div`
  color: ${({ theme }) => theme.color.text2};

  font-size: 14px;
  /** checkbox の height が 20px なのでcheckbox と text が揃っているように見せるために行ボックスの高さを 20px にしている */
  line-height: 20px;
`
