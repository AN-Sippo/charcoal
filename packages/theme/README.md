# `@charcoal/theme`

charcoal 全体で制約されるテーマ構造の型を提供します。

## インストール

npm

```
npm i --save-dev @charcoal/theme
```

yarn

```
yarn add -D @charcoal/theme
```
