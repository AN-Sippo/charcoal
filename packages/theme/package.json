{
  "name": "@charcoal-ui/theme",
  "version": "4.0.0-beta.15",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    "types": "./dist/index.d.ts",
    "require": "./dist/index.cjs.js",
    "import": "./dist/index.esm.js",
    "default": "./dist/index.esm.js"
  },
  "types": "./dist/index.d.ts",
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "npm-run-all --print-label --parallel 'build:*' --sequential serialize",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "build:copy-css": "cpx 'src/css/**/*.css' dist/css && cpx 'src/json/**/*.json' dist/json",
    "serialize": "node cli/index.js",
    "typecheck": "run-p --print-label 'typecheck:*'",
    "typecheck:main": "tsc --project tsconfig.build.json --pretty --noEmit",
    "typecheck:cli": "tsc --project cli/tsconfig.build.json --noEmit",
    "clean": "rimraf dist .tsbuildinfo",
    "test": "vitest run --passWithNoTests"
  },
  "devDependencies": {
    "cpx": "^1.5.0",
    "npm-run-all": "^4.1.5",
    "rimraf": "^3.0.2",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5",
    "vitest": "^2.0.2"
  },
  "dependencies": {
    "@charcoal-ui/foundation": "^4.0.0-beta.15",
    "@charcoal-ui/utils": "^4.0.0-beta.15",
    "polished": "^4.1.4"
  },
  "files": [
    "src",
    "dist",
    "theme.json"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/theme"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
