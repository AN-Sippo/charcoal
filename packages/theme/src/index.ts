import { Effect, Effects, OpacityEffect } from './effect'

export default {}

export * from './effect'

export type Material = string
export type GradientMaterial = { color: Material; ratio: number }[]

/**
 * @deprecated
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type ThemeMaterials = {
  background1: Material
  background2: Material
  surface1: Material
  surface2: Material
  /**
   * Black Fade 20にするとボーダーと共通化できる
   */
  surface3: Material
  surface4: Material
  // NOTE: the current definition of surface5 is invalid
  surface6: Material
  /**
   * サムネイルに被せる色。
   * Lightモードの色が2と共通。コントラストの低下を狙うなら2でいいが、サムネイルで色味を調整しすぎる感がある。
   */
  surface7: Material
  /**
   * 磨りガラス効果を使えれば6と統合できそう
   */
  surface8: Material
  surface9: Material
  link1: Material
  link2: Material
  text1: Material
  /**
   * 1と2は統合できるかもしれない
   */
  text2: Material
  text3: Material
  /**
   * Surface 4に重ねた時に薄すぎる感あり
   */
  text4: Material
  text5: Material
  icon6: Material
  border: Material
}

/**
 * @deprecated
 */
export interface ThemeEffects {
  hover: Effects
  press: Effects
  disabled: Effects
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeEffect = {
  hover: Effect
  press: Effect
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeElementEffect = {
  disabled: OpacityEffect
}

/**
 * @deprecated
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type ThemeSemantic = {
  brand: Material
}

/**
 * @deprecated Use `ThemeColor` instead
 */
export type ThemeColors = ThemeMaterials &
  ThemeSemantic & {
    premium: Material
  }

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeColor = {
  transparent: Material
  background1: Material
  background2: Material
  surface1: Material
  surface2: Material
  surface3: Material
  surface4: Material
  surface6: Material
  surface7: Material
  surface8: Material
  surface9: Material
  link1: Material
  link2: Material
  text1: Material
  text2: Material
  text3: Material
  text4: Material
  text5: Material
  icon6: Material
  border: Material
  brand: Material
  assertive: Material
  /**
   * @deprecated
   */
  premium: Material
}

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type ThemeColorGradient = {
  surface4Gradient: GradientMaterial
}

export interface Theme {
  /**
   * @deprecated
   *
   * Use `color` instead.
   */
  materials: ThemeMaterials
  /**
   * @deprecated
   *
   * Use `color` instead.
   */
  semantic: ThemeSemantic
  /**
   * @deprecated Use `color` instead
   */
  colors: ThemeColors
  color: ThemeColor
  gradientColor: ThemeColorGradient
  /**
   * @deprecated Use `effect` instead
   */
  effects: ThemeEffects
  /**
   * effect for color (background or font color)
   */
  effect: ThemeEffect
  /**
   * effect for element its own (opacify element includes descendants)
   */
  elementEffect: ThemeElementEffect
}
