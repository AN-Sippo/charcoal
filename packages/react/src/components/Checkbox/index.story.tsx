import { action } from '@storybook/addon-actions'
import Checkbox from '.'
import { Meta, StoryObj } from '@storybook/react'
import { useCallback, useState } from 'react'

export default {
  title: 'Checkbox',
  component: Checkbox,
} as Meta<typeof Checkbox>

export const Default: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
    invalid: false,
  },
  render: (props) => {
    const [checked, setChecked] = useState(props.checked)
    const handleChange = useCallback((isSelected: boolean) => {
      setChecked(isSelected)
      action('change')(isSelected)
    }, [])

    return (
      <div>
        <Checkbox
          {...props}
          checked={checked ?? props.checked}
          name="labelled"
          label="label"
          onBlur={action('blur')}
          onClick={action('click')}
          onChange={handleChange}
          onFocus={action('focus')}
        >
          <span style={{ userSelect: 'none' }}>同意する</span>
        </Checkbox>
      </div>
    )
  },
}

export const Invalid: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
  },
  render: (props) => {
    return (
      <Checkbox
        {...props}
        name="labelled"
        label="label"
        invalid
        onBlur={action('blur')}
        onClick={action('click')}
        onChange={action('change')}
        onFocus={action('focus')}
      >
        同意する
      </Checkbox>
    )
  },
}

export const Unlabelled: StoryObj<typeof Checkbox> = {
  args: {
    checked: false,
    defaultChecked: false,
    disabled: false,
    readonly: false,
    invalid: false,
  },
  render: (props) => {
    return (
      <div>
        <Checkbox
          {...props}
          name="unlabelled"
          label="label"
          onBlur={action('blur')}
          onClick={action('click')}
          onChange={action('change')}
          onFocus={action('focus')}
        />
      </div>
    )
  },
}
