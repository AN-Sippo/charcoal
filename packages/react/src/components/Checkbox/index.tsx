import './index.css'

import React, { forwardRef, memo } from 'react'
import { useId } from '@react-aria/utils'
import CheckboxInput, { CheckboxInputProps } from '../CheckboxInput'
import { useClassNames } from '../../_lib/useClassNames'

export type CheckboxProps = CheckboxInputProps & {
  children?: React.ReactNode
}

const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  function CheckboxInner({ disabled, className, id, children, ...props }, ref) {
    const classNames = useClassNames('charcoal-checkbox__label', className)
    const htmlId = useId(id)
    const input = (
      <CheckboxInput {...props} id={htmlId} disabled={disabled} ref={ref} />
    )
    if (children === undefined) {
      return input
    }
    return (
      <label htmlFor={htmlId} aria-disabled={disabled} className={classNames}>
        {input}
        <div className="charcoal-checkbox__text">{children}</div>
      </label>
    )
  }
)

export default memo(Checkbox)
