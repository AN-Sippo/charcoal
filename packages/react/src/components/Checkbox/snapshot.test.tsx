import { dark, light } from '@charcoal-ui/theme'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import Checkbox from '.'

function render(children: JSX.Element, theme = light) {
  return renderer
    .create(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
    .toJSON()
}

describe('Checkbox Snapshot Test', () => {
  it('light theme snapshot', () => {
    expect(render(<Checkbox label="test" />, light)).toMatchSnapshot()
  })
  it('dark theme snapshot', () => {
    expect(render(<Checkbox label="test" />, dark)).toMatchSnapshot()
  })
})
