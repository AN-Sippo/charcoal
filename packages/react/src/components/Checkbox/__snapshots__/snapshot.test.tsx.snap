// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Snapshot Test dark theme snapshot 1`] = `
<label
  aria-disabled={false}
  className="sc-bdvvtL bWvKyY"
>
  <div
    className="sc-gsDKAQ dFzOKQ"
  >
    <input
      aria-label="test"
      checked={false}
      className="sc-dkPtRN lfMjCb"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="checkbox"
    />
    <div
      aria-hidden={true}
      checked={false}
      className="sc-hKwDye hoolJa"
    >
      <pixiv-icon
        name="24/Check"
        unsafe-non-guideline-scale={0.6666666666666666}
      />
    </div>
  </div>
</label>
`;

exports[`Checkbox Snapshot Test light theme snapshot 1`] = `
<label
  aria-disabled={false}
  className="sc-bdvvtL bWvKyY"
>
  <div
    className="sc-gsDKAQ dFzOKQ"
  >
    <input
      aria-label="test"
      checked={false}
      className="sc-dkPtRN jYXIQQ"
      disabled={false}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="checkbox"
    />
    <div
      aria-hidden={true}
      checked={false}
      className="sc-hKwDye hoolJa"
    >
      <pixiv-icon
        name="24/Check"
        unsafe-non-guideline-scale={0.6666666666666666}
      />
    </div>
  </div>
</label>
`;
